src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -0,0 +1,107 @@
+package org.sagebionetworks.bridge.services;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.sagebionetworks.bridge.validators.PermissionValidator.INSTANCE;
+
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.dao.PermissionDao;
+import org.sagebionetworks.bridge.exceptions.EntityNotFoundException;
+import org.sagebionetworks.bridge.models.accounts.Account;
+import org.sagebionetworks.bridge.models.accounts.AccountId;
+import org.sagebionetworks.bridge.models.accounts.AccountRef;
+import org.sagebionetworks.bridge.models.permissions.EntityType;
+import org.sagebionetworks.bridge.models.permissions.Permission;
+import org.sagebionetworks.bridge.models.permissions.PermissionDetail;
+import org.sagebionetworks.bridge.validators.Validate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@Component
+public class PermissionService {
+    
+    private PermissionDao permissionDao;
+    
+    private AccountService accountService;
+    
+    @Autowired
+    final void setPermissionDao(PermissionDao permissionDao) {
+        this.permissionDao = permissionDao;
+    }
+    
+    @Autowired
+    final void setAccountService(AccountService accountService) {
+        this.accountService = accountService;
+    }
+    
+    // For mock testing
+    protected String generateGuid() {
+        return BridgeUtils.generateGuid();
+    }
+    
+    public PermissionDetail createPermission(String appId, Permission permission) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(permission);
+        
+        permission.setGuid(generateGuid());
+        
+        Validate.entityThrowingException(INSTANCE, permission);
+        
+        Permission createdPermission = permissionDao.createPermission(appId, permission);
```

Comment:
If the user or entity doesn't exist, what error does it throw? Status code? Is the error message reasonable?

Reply:
Added exception handling through the MySQLHibernatePersistenceExceptionConverter to handle foreign keys to non-existent entities. It throws a 409 ConstraintViolationException now. The message reads: "This permission cannot be created or updated because the referenced user account does not exist." (user account is replaced with study, assessment, or organization if the foreign key issue was based on the entityId)